python/pyspark/sql/worker/plan_data_source_read.py

```diff
@@ -51,6 +52,71 @@
 )
 
 
+def records_to_arrow_batches(
+    output_iter: Iterator[Tuple],
+    max_arrow_batch_size: int,
+    return_type: StructType,
+    data_source: DataSource,
+) -> Iterable[pa.RecordBatch]:
+    def batched(iterator: Iterator, n: int) -> Iterator:
+        return iter(functools.partial(lambda it: list(islice(it, n)), iterator), [])
```

Comment:
Would the batches be constructed immediately here? Instead, can we do lazy generation using generators?

Reply:
This part of code was refactored out of the old plan_data_source_read file, I am not sure whether this is lazy or not. But for streaming we assume that simple python data source is light weight without partitioning and can materialize all records of a microbatch at once, so it doesn't matter.